modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx

```diff
@@ -158,49 +214,142 @@ const AddFieldsModalContent = ({
 
 				setSelectedKeys(initialSelectedKeys);
 
-				setFields(updatedFields);
+				setInitialFields(updatedFields);
 			}
 		});
 	}, [savedFDSFields, fdsView]);
 
+	useEffect(() => {
+		if (initialFields) {
+			const filteredItems = filterItems(initialFields, debouncedSearch);
+			setFields(filteredItems);
+
+			if (debouncedSearch) {
+				const filteredKeys = getItemsKeys(filteredItems);
+				setExpandedKeys(filteredKeys);
+			}
+		}
+	}, [initialFields, debouncedSearch]);
```

Comment:
Do we need `useEffect` here? It looks to me like the dependencies will never change. If that's the case, we should be able to set initial setting of states in `useState`s.

It's typically wise to avoid `useEffect` whenever possible:
- It adds additional re-render, which is a small hit on performance.
- Additional re-render can lead to bugs and glitchy behavior 
- "event bus" nature of `useEffect` ends up in code that's often hard to debug.

Reply:
I'm not sure, the `debouncedSearch` is updated when the user type.

Reply:
Oh, my bad! I misread the code, I though `debouncedSearch` was a function. Both `debounce` util in DXP and the old `A.debounce` return a function :sweat_smile: 

But, my point with avoiding `useEffect` is still valid. Is there a way to refactor code to directly modify `fields` and `expandedKeys` states in search input callback?